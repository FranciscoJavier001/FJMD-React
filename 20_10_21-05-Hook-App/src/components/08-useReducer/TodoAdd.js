import React from 'react'
import { useForm } from '../../hooks/useForm';

// Exportamos el TodoAdd a TodoApp
export const TodoAdd = ({ handleAddTodo }) => { 

    // Este es el estado del formulario 
    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    // Va a ser igual a una funcion que recibe el event 
    const handleSubmit = (e) => { 
        e.preventDefault(); //** Con eso detengo la propagacion del formulario que se envia por el handlesubmit*/

        // Si la tarea es menor o igual a una no hagas nada
        if ( description.trim().length <= 1 ) { 
            return;
        }

        //** Creado el newTodo con la descripcion  */
        const newTodo = {
            id: new Date().getTime(), //** milisegundos */
            desc: description,
            done: false
        };
        handleAddTodo( newTodo ); //** Este lo va a llamar mandando el newTodo, no voy a mandar la accion, voy a mandar el newTodo */
        reset(); //** El reset del formulario es el que limpia el campo del imput donde dice aprender */
    }

    return (
        // fragment
        <> 
        {/* Titulo */}
            <h4>Agregar Tarea</h4> 
            {/* linea */}
            <hr /> 

            {/* El formulario tiene un submit, que al apretarle va a mandar el formulario, pero sin que la pagina se recarge */}
            <form onSubmit={ handleSubmit }>  

                <input //** Este es el campo de texto */
                    type="text"
                    name="description"
                    className="form-control" //** Para que use todo el espacio */
                    placeholder="Ingresa alguna Actividad ..."
                    autoComplete="off"
                    value={ description } //** Esto significa que se limpie el input */
                    onChange={ handleInputChange } //** Esto bloquea para que no se pueda escribir en el input  */
                />

                <button
                    type="submit" //** Tipo de boton */
                    className="btn btn-outline-primary mt-1 btn-block" //** Clase bootstrap */
                >
                    {/* Como se va a llamar */}
                    Agregar
                </button>
            </form>           
        </>
    )
}