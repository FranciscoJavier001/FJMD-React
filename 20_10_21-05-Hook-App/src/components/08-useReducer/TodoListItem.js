import React from 'react';

// Estas son las propTypes que tiene que recibir
export const TodoListItem = ({ todo, index, handleDelete, handleToggle }) => { 

    return (
        <li
            key={ todo.id } //** Aqui creamos el key que va a ser el todo.id, que importamos en el TodoList */
            className="list-group-item" //** Clase de bootstrap que hace una lista de grupal */
        >  
            <p 
                className={ `${ todo.done && 'complete' }` } //** Aqui hicimos la desestructuracion de archivos, que nos manda la clase del css, donde pusimos si el todo.done &&  */
                onClick={ () => handleToggle( todo.id ) } //** Es un manejador de eventos */
            > 
            {/* El primer index indica de donde va a empezar y luego el punto es lo que sigue del numero, luego la descripcion del todo*/}
                { index + 1}. { todo.desc }   
            </p>
            <button
                className="btn btn-danger"
                onClick={ () => handleDelete( todo.id ) } //** Manejador de eventos cuando haga click en el boton borrar */
            >
                Borrar
            </button>
        </li>
    )
}
