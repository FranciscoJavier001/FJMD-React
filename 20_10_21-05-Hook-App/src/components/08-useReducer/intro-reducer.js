
const initialState = [{ //** Inicializamos el initialState, pero no importa lo que hagamos */
    id: 1,
    tarea: 'Comprar pan',
    lista: false
}];

const listaTareas = ( state = initialState, action ) => { //** Funcion de flecha que recibe el state como primer parametro que esre es igual al initial state */
    
    if ( action?.type === 'agregar' ) { //** El sign de interrogacion nos dice uue si no recibe un parametro no haga nada */
        return [ ...state, action.payload ];
    }

    return state;
}

let tareas = listaTareas();

const nuevaTarea = {
    id: 2,
    tarea: 'Comprar leche',
    lista: false
}

const agregarNuevaTarea = {
    type: 'agregar',
    payload: nuevaTarea
}

tareas = listaTareas( tareas, agregarNuevaTarea );

console.log(tareas);

