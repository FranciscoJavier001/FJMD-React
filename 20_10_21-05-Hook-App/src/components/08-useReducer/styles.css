

body { 
    padding: 50px; /*Es el Relleno que tiene el espacio*/
}

/* Este es el numero de la lista y la tarea */
p {
    cursor: pointer;
    margin: 0px;
}

/* Aqui es donde se muestra la lista de tareas, pero esto es el boton */
.list-group-item { 
    align-items: center;
    display: flex;
    justify-content: space-between;
}

/* Si esta esta en true, entonces ponle una linea que pase por enmedio y retorna la clase complete */
.complete { 
    text-decoration: line-through;
}