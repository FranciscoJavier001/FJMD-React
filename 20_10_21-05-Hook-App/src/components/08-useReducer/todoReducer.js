//** Aqui creamos la funcion todoReducer, que la van a importar en el TodoApp */

//** todoReducer tiene una funcion que recibe parametros, el state es el arreglo actual y el action sigue las instrucciones que viene a continuacion en el todoApp */
export const todoReducer = ( state = [], action ) => { 


    // Crear las acciones y mandarselas al reducer, el reducer las ejecuta, me regresa un nuevo estado y como estoy trabajando en base a mi useReducer este se va a encargar de redibujar lo que hay que redibujar 
    switch ( action.type ) { /** Wey, esto ya es el boton, con toda la logica para que agrege una nueva tarea */
        case 'add':
            return [ ...state, action.payload ]; /** Como nos damos cuenta, aqui se va a agregar, y la ltima va a ser la primera, manteniendo la fotografia como estaba antes*/

        case 'delete': //** En caso de que el usuario en el TodoApp *///** Este el de borrar, y vamos a hacer un filter que va a regresar un nuevo arreglo del state que cumplan una condicion */

            return state.filter( todo => todo.id !== action.payload ); // 123123123 //** Esto nos dice que el Todo sea diferente al todo.id que quiero borrar, que sea diferente al action.payload (este es donde se guardan las cosas que se van a cambiar porque este cambio el estado a uno nuevo con el payload), mira hay te va mejor la sentencia - En cas*/

            //** es mas sensillo hacer una funcion que tenga un return implicito con un operador ternario, recuerden que el ternario tiene que regresar dos valores que es lo que se necesita uno que seria la modificacion y el otro en caso de que sea igual y no quiera hacer nada. */

        case 'toggle':
            return state.map( todo => //** map barre cada uno de los elementos del arreglo que seria el state y debe de retornar un valor, lo que sea que retorne la funcion o el callback de aqui dentro sera el nuevo valor que va a tener cada uno de los elementos de ese state - aqio vo a tener un todo*/
                ( todo.id === action.payload ) //** action.payload va a venir el id del mismo - si el todo.id es exactamente igual al accion.payload significa que hay que cambiar este elemento  */
                    ? { ...todo, done: !todo.done } //** el operador spreed, que me regresa el arreglo como esta actualmente pero el cambio se viene hasta abajo y lo unico que me interesa cambiar es el todo.done la propiedad done va a tener el valor del todo.done pero quiero hacer la negacion para que sean diferentes */
                    : todo //** si son diferentes solamente voy a retornar el todo */
            );
        
                    case 'toggle-old': //** map barre cada uno de los elementos del arreglo que seria el state y debe de retornar un valor, lo que sea que retorne la funcion o el callback de aqui adentro sera el nuevo valor que va atener cada uno de los elementos de ese state  */
            return state.map( todo => {//** aqui voy a tener un todo  */

                if ( todo.id === action.payload ) { //** action.payload va a venir el id del mismo - si el todo.id es exactamente igual al accion.payload significa que hay que cambiar este elemento  */
                    return {
                        ...todo, done: !todo.done} //** la propiedad done va a tener el valor del todo.done pero quiero hacer la negacion para que sean diferentes */
                } else {
                    return todo; //** en caso contrario no son iguales yse va hacer un retorn del mismo todo  para no mutar y no estruir algo, porque si no regresaramos nada en el mapa entonces va atransformar el todo en un undefine, por eso hay que regresar algo ahi */
                }
            })

        default:
            return state; /** Si vemos que no hay cambio manda el mismo estado */
    }


}