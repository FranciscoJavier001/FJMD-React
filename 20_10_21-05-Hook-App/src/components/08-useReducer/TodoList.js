import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => { //** Exportamos TodoList, que esta en TodoApp y le asiganos un valor a cada una de las variables, pero las pasamos ya desestructuradas, estas 3 variables ya estan en TodoApp */
    return ( //** Vamos a retornar */

        // list-group crea una lista, que se mueve hacia la izquierda, y el list-group-flush elimina el recuadro del input
         <ul className="list-group list-group-flush"> 
        {
            todos.map( (todo, i) => (
                <TodoListItem //** Este es el componente, que importamos del TodoListItem */
                    key={ todo.id } //** Tenemos que definir todos estos, el indice nunca hay que usarlo como key */
                    todo={ todo }
                    index={ i }
                    handleDelete={ handleDelete } //** Es la referencia a la funcion */
                    handleToggle={ handleToggle } //** Es la referencia a la funcion */
                />
            ))
        }
    </ul>
    )
}
