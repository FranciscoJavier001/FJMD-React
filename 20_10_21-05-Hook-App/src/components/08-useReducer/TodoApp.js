//******************************/ dispatch signific envio
// Imporaciones que se pueden mandar llamar desde aqui

import React, { useReducer, useEffect } from 'react';
import { todoReducer } from './todoReducer';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';

//** Funcion de flecha que no recibe parametros, pero si las instrucciones que vienen despues, que establece el nuevo estado de los todos en el todoReducer */
const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || []; //** Analiza un objeto, y lo transforma en el estado actual, se puso arriba porque es una funcion que se usa en el useEffect (que esta quiere decir "haz el cambio"), que ya se definio abajo y si esta null no regresa nada (pero creo que eso es mas para evitar conflictos), el JSON.parse nos devuelve lo wue tiene dentro el arreglo, por so pusimos que en caso de or nos devuelva el arreglo vacio */
}

// Esto se importa desde todoReducer, estas son las acciones que viene en el Case
export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init); //** El useReducer, maneja el todoReducer que aqui lo importamos, maneja un arreglo con el state actual, y que tiene como estado inicial el init, y los todos que son las acciones actuales y el dispatch, que son las acciones hacia mi reducer /*

    // Va a disparar un callback (que es una funcion que recibe como argumento otra funcion y la ejecuta), si los todos cambian eso significa que vamos a volver a ejecutar la funcion del localStorage y eso es lo que hace el useEffect 
    useEffect( ()=> {
        localStorage.setItem('todos', JSON.stringify( todos ) ); /** Este va a guadar en el LocalStorage cualquier cambio que pase en los todos de la constante de abajo, y el JSON.stringyfy los convierte en strings */ 
    }, [todos]); //** Como los todos cambiaron, entonces se actualizan y vuelve a iniciar la instruccion */

    
    //** Estas son funciones constantes que van a recibir un objeto con sus argumentos y puede generar las siguientes acciones que van a ser guardadas en el todoId */
    const handleDelete = ( todoId ) => {
        const action = { //** action se declara aqui, pero se va a acceder a ella desde el todoReducer en el switch, y si algo cambio en automatico se va a ver reflejado, porque esta comparado con el action.payload que se guarda el cambio en el todoId
            type: 'delete', 
            payload: todoId 
        }
        dispatch( action ); /** Aqui mandamos la nueva accion que se hizo arriba pero ya guardada y lanzada pero ahora como el init nuevo! */
    }

    const handleToggle = ( todoId ) =>{ //** Recibe el todoId que es donde quiero hacer el cambio y ponerle la linea*/
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }
//******************************/
// crear nueva funcion que se llame handleAddTodo que sea una funcion de flecha que va a recibir un nuevo todo que deberia ser un objeto 
    const handleAddTodo = ( newTodo ) => { 
        dispatch({ //** Este es el objeto que crea el action  */
            type: 'add',
            payload: newTodo
        });
    }

    // Aqui retornamos el numero de arrelos del Todo
    return ( 
        <div>
            <h1>TodoApp ( { todos.length } ) </h1>  {/* Aqui es el length que esta en el titulo que cabia segun los todos que tenga inscritos */}
            <hr /> {/* Linea */}

            {/* Estas con las clases de bootstrap */}
            <div className="row"> 
                <div className="col-7">

                    {/* Aqui es donde estamos a la espera de que sea clickeado cualquier boton y si  */}
                    <TodoList //** Este es el nombre del componente (Asi se declaran los componentes y los componentes se pueden exportar con la palabra export y se pueden importr con la palabra import ) */
                        todos={ todos } //** Estas son las referencias, pero para que? */
                        handleDelete={ handleDelete }  //** Estas son las referencias */
                        handleToggle={ handleToggle } //** Estas son las referencias  */
                    />
                </div>

                {/* Aqui es donde se muestra lo que se va a agregar, es el input y el boton */}
                {/* Clase bootstrap */}
                <div className="col-5"> 
                    <TodoAdd //** Agregamos el componente de TodoAdd  */
                        handleAddTodo={ handleAddTodo } //** Estas son las referencias a la funcion, por consecuencia mi componente de todoAdd va a recibir la funcion ahi, para que pase por el reducer */
                    />
                </div>
                
            </div>
        </div>
    )
}