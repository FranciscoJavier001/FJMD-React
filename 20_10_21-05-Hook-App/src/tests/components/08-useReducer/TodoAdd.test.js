import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Pruebas en <TodoAdd />', () => {

    const handleAddTodo = jest.fn(); //** Mook de la funcion, creamos la referencia a la funcion */

    const wrapper = shallow( //** Creamos la constanste wrapper igual a shallow e importamos el componente del todoAdd, este componente necesita una funcion, hacemos el mook de la funcion en la linea 7, necesitamos el handleAddTodo y que va a ser igual al handleAddTodo */
        <TodoAdd 
            handleAddTodo={ handleAddTodo }
        /> 
    )


    test('debe de mostrarse correctamente', () => { //** Hacemos un match con el wrapper */
        
        expect( wrapper ).toMatchSnapshot(); //** Que haga match con el snapshot para segurarnos que el componente no va a cambiar dr manera inexperada */
    });

    test('NO debe de llamar handleAddTodo', () => { //** Como no hay ningun value en la descripcion no se va a mandar porque no tiene nada */
        
        const formSubmit = wrapper.find('form').prop('onSubmit'); //** Aqui buscamos el form con el find usamos el property que va a funcionar para cualquier libreria de terceros y buscamos la property que sea el onSubmit que hace referencia a una funcion */

        formSubmit({ preventDefault(){} }); //** Asi llamamos una funcion, pero devemos de mandar el preventDefault para que no reviente y para decir que es una funcion le ponemos el {} */

        expect( handleAddTodo ).toHaveBeenCalledTimes(0); //** Espero que el handleAddTodo tenga que haber sido llamado 0 veces */
    });

    test('debe de llamar la funciÃ³n handleAddTodo', () => {
        
        const value = 'Aprender Firestore'; //** Primero el value que va a tener la caja de texto */
        wrapper.find('input').simulate('change', { //** Que lo busque en el input que tenemos y vamos a simular el change y el argumento va a ser el target que va a tener el value y el name y el nombre va a ser la descripcion que es el campo que quiero cambiar */
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit'); //** Tenemos la referencia al formulario y deberia de haberse llamado 1 vez, osea que si tuvimos un valor y que lo quisimos cambiar en el handleAddTodo */
        formSubmit({ preventDefault(){} });

        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) ); // { } //** Necesitamos que sea llamado conra un objeto, que es el que vamos a evaluar, asi podriamos esperar cualquier objeto */
        expect( handleAddTodo ).toHaveBeenCalledWith({ //** Asi hacemos la prueba mas precisa */
            id: expect.any(Number), //** Asi definimos el objeto, pero ponemos que el numero sea any */
            desc: value,
            done: false
        }); 

        expect( wrapper.find('input').prop('value') ).toBe(''); //** Asi comprobamos que haya mandado llamar al reset de la funcion, entonces hacemos un expect, el efecto que tiene la funcion el value tiene como valor un string vacio, buscamos el input, vamos a evaluar la property value y esta debe ser un string vacio */
    })
})
