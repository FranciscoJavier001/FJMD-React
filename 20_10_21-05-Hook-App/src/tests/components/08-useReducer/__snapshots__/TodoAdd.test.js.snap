// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar Tarea
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Ingresa alguna Actividad ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
